File: packages/sdk-flutter/lib/src/scorecard_evaluator.dart
import "models.dart";

class ScorecardEvaluator {
  Map<String, dynamic> evaluate(Scorecard scorecard, Map<String, dynamic> variables) {
    var total = scorecard.baseScore;
    final breakdown = <Map<String, dynamic>>[];
    for (final factor in scorecard.bins) {
      final value = _numeric(variables[factor.variableId]);
      ScoreRange? activeRange;
      for (final range in factor.ranges) {
        final min = range.min ?? double.negativeInfinity;
        final max = range.max ?? double.infinity;
        if (value >= min && value <= max) {
          activeRange = range;
          break;
        }
      }
      final points = activeRange?.points ?? 0;
      total += points;
      breakdown.add(<String, dynamic>{
        "variable_id": factor.variableId,
        "value": value,
        "points": points,
        "active_range": activeRange == null
            ? null
            : <String, dynamic>{
                "min": activeRange.min,
                "max": activeRange.max,
                "points": activeRange.points,
              },
      });
    }
    return <String, dynamic>{
      "score": total.clamp(0, scorecard.maxScore),
      "base_score": scorecard.baseScore,
      "max_score": scorecard.maxScore,
      "breakdown": breakdown,
    };
  }

  double _numeric(dynamic value) {
    if (value is num) {
      return value.toDouble();
    }
    if (value is String) {
      return double.tryParse(value) ?? 0.0;
    }
    return 0.0;
  }
}

File: packages/sdk-android/rulemind-core/src/main/kotlin/com/rulemind/core/ScorecardEvaluator.kt
package com.rulemind.core

import com.rulemind.core.models.Scorecard

class ScorecardEvaluator {
    fun evaluate(scorecard: Scorecard, variables: Map<String, Any?>): Map<String, Any?> {
        var total = scorecard.baseScore
        val breakdown = mutableListOf<Map<String, Any?>>()
        for (factor in scorecard.bins) {
            val value = when (val raw = variables[factor.variableId]) {
                is Number -> raw.toDouble()
                is String -> raw.toDoubleOrNull() ?: 0.0
                else -> 0.0
            }
            val activeRange = factor.ranges.firstOrNull { range ->
                val min = range.min ?: Double.NEGATIVE_INFINITY
                val max = range.max ?: Double.POSITIVE_INFINITY
                value >= min && value <= max
            }
            val points = activeRange?.points ?: 0
            total += points
            breakdown += mapOf(
                "variable_id" to factor.variableId,
                "value" to value,
                "points" to points,
                "active_range" to activeRange?.let {
                    mapOf(
                        "min" to it.min,
                        "max" to it.max,
                        "points" to it.points,
                    )
                },
            )
        }
        return mapOf(
            "score" to total.coerceIn(0, scorecard.maxScore),
            "base_score" to scorecard.baseScore,
            "max_score" to scorecard.maxScore,
            "breakdown" to breakdown,
        )
    }
}
